JavaScript/ContactBook/app.js

```diff
@@ -0,0 +1,338 @@
+import onChange from "https://cdn.jsdelivr.net/npm/on-change@5.0.1/+esm";
+import { uid } from "https://cdn.jsdelivr.net/npm/uid@2.0.2/+esm";
+
+/**
+ * Завдання 3:
+ * Реалізуйте інтерфейс для відображення та управління контактами з телефонної книги з можливістю:
+ * Додавання нових контактів через форму.
+ * Редагування існуючих контактів.
+ * Видалення контактів.
+ * Пошуку контактів за ім'ям або номером телефону.
+ *
+ * Вимоги:
+ * Використовуйте методи document.createElement, appendChild, addEventListener для створення елементів телефонної книги та управління ними.
+ * Забезпечте динамічне оновлення списку контактів при додаванні, редагуванні та видаленні.
+ * Зберігайте контакти в LocalStorage, щоб вони зберігались між перезавантаженнями сторінки.
+ *
+ * Підказка:
+ * Створіть структуру телефонної книги з використанням HTML-елементів ul та li.
+ * Додайте форми для додавання та редагування контактів, а також поле для пошуку.
+ */
+
+function app(containerId) {
+  const initialState = {
+    contacts: [],
+    uiState: {
+      editId: null,
+      searchQuery: "",
+    },
+  };
+
+  const containers = initLayout(containerId);
+  const watchedState = onChange(initialState, render(containers));
+  const storedContacts = JSON.parse(localStorage.getItem("contacts")) || [];
+
+  // initialising tasks and fires first rerendering of the app
+  watchedState.contacts = storedContacts;
```

Comment:
Review if there's a need to reinitialize contacts on every app initialization or if we can optimize it.